Add section about syntax special characters

Hunterness opened the pull request:
We're using a custom syntax for describing the administration commands but we never explain what our special characters mean, for example `{ ... }` or `[ ... ]`.

Hunterness committed:
Update relationship constraint syntax

So that the `[` in the pattern is not confused with the `[` indicating optional

Hunterness committed:
Add section about syntax special characters

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +65,69 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+Some things to notice about this command is that it includes `{` and `}` in the syntax, and between them has a grouping of either a list of properties or the character `*`.
+It also has multiple optional parts, including the entity part of the command which is the grouping following the graph name.
+
+In difference, there is no need to escape any characters in the index creation command.
+This since `(` and `)` are not special characters and `[` and `]` are not part of the command, but indicate that the index name is optional.
+
+[source, cypher]
+----
+CREATE INDEX [index_name]
+FOR (n:LabelName)
+ON (n.propertyName)
+----
```
Change to node property existence constraint to avoid `OPTIONS` clause needing escaping in 4.2

davidoliverSP2 committed:
editorial update

davidoliverSP2 reviewed (approved):
Thanks @Hunterness 

I pushed some editorial changes, and will wait for the builds to go green. I'm aware that this edits will need to be applied to the PRs in later versions also 😄

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -46,6 +46,7 @@ This section explains how to use Cypher to manage Neo4j role-based access contro
 ** <<administration-security-limitations-labels, Security and labels>>
 ** <<administration-security-limitations-db-operations, Security and count store operations>>
 
+
```
This is very inconsistent in all our administration files, some places have 2 and some have 1 blank lines, would probably benefit from going an overall sweep and make it consistent

davidoliverSP2 replied:
Yep, we actually have review + restructure this section as an action on our wall - we are just looking for an opportune moment to do it, as it would likely take us around 2-3 weeks and would need to freeze any submissions while we do it! Maybe in scope for the last few weeks of the year though

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -65,28 +66,34 @@ Instead they are defined in terms of their underlying _privileges_ and they can
 
 .Special characters in syntax summeries
-[options="header", width="100%", cols="1a,3,3"]
+[options="header", width="100%", cols="1a^,3a,3a"]
```
Yeah I have no real clue what these numbers and letters mean :P

davidoliverSP2 replied:
😄  this is asciidoc syntax for the table layout - nothing that we'd expect you to add to your PRs, and is something we're happy to pick up in reviews.

The `1a^` means column width of 1, and that content within is asciidoc (`a`)markup, and to center the text (`^`).
The `3a` means the columns should have width of 3, and also have asciidoc markup. 

Here is a useful guide: https://blog.mrhaki.com/2014/11/awesome-asciidoctor-styling-columns-and.html

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -99,10 +106,11 @@ The special characters and their meaning are as follows:
 |===
 
-Any other characters do not have any special meaning, therefore not needing to be escaped.
+There are no other characters that need to be escaped.
```
This still have @Lojjs 'problem' with using escaped in a completely different way then what we normally use in cypher (where things are escaped using `)

It's one of her comments on the 4.3 version

davidoliverSP2 replied:
ok, will look ahead at those review comments in #758

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +66,76 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+| `\|`
+| Or; used to indicate alternative parts of a command.
+Needs to be part of a grouping.
+| If the syntax needs to specify either a name or `*`, this can be indicated with `name \| *`.
```
```suggestion
| If the syntax needs to specify either a name or `*`, this can be indicated with `* \| name`.
```
To make the order more consistent (have a card to update other places)

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +66,76 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+
+| `{` and `}`
+| Used to group parts of the command, common together with `\|`.
+| To use the `or` in the syntax summary it needs to be in a group, `{name \| *}`.
```
```suggestion
| To use the `or` in the syntax summary it needs to be in a group, `{* \| name}`.
```

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +66,76 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+
+| `"`
+| When a special character is part of the syntax itself, we surround it with `"` to indicate this.
+| To include `{` in the syntax use `"{" { name \| * } "}"`, here we get either `{*}` or `{name}`.
```
```suggestion
| To include `{` in the syntax use `"{" { * \| name } "}"`, here we get either `{*}` or `{name}`.
```

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +66,76 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+----
+GRANT READ
+    "{" { * | property[, ...] } "}"
+    ON GRAPH[S] {name | *}
```
```suggestion
    ON GRAPH[S] {* | name}
```

Hunterness committed:
Change order of name and *

To make the order more consistent in the Administration chapter (have a card to update other places)

Hunterness commented on cypher/cypher-docs/src/docs/dev/ql/administration/security/index.asciidoc:
```diff
@@ -64,6 +66,76 @@ Instead they are defined in terms of their underlying _privileges_ and they can
+----
+GRANT READ
+    "{" { * | property[, ...] } "}"
+    ON GRAPH[S] {* | name}
```
```suggestion
    ON GRAPH[S] { * | name }
```

Hunterness committed:
Add spaces for consistency
